source/fuzz/transformation_duplicate_region_with_selection.cpp

```diff
@@ -0,0 +1,563 @@
+      result.insert(&block);
+    }
+  }
+  return result;
+}
+
+void TransformationDuplicateRegionWithSelection::Apply(
+    opt::IRContext* ir_context, TransformationContext* /*unused*/) const {
+  fuzzerutil::UpdateModuleIdBound(ir_context, message_.new_entry_fresh_id());
+  fuzzerutil::UpdateModuleIdBound(ir_context, message_.merge_label_fresh_id());
+
+  // Create the new entry block containing the main conditional instruction. Set
+  // its parent to the parent of the original entry block, since it is located
+  // in the same function.
+  std::unique_ptr<opt::BasicBlock> new_entry_block =
+      MakeUnique<opt::BasicBlock>(MakeUnique<opt::Instruction>(
+          ir_context, SpvOpLabel, 0, message_.new_entry_fresh_id(),
+          opt::Instruction::OperandList()));
+  auto entry_block = ir_context->cfg()->block(message_.entry_block_id());
+  auto enclosing_function = entry_block->GetParent();
+  new_entry_block->SetParent(enclosing_function);
+  auto exit_block = ir_context->cfg()->block(message_.exit_block_id());
+
+  // Get the blocks contained in the region.
+  std::set<opt::BasicBlock*> region_blocks =
+      GetRegionBlocks(ir_context, entry_block, exit_block);
+
+  // Construct the merge block and set its parent.
+  std::unique_ptr<opt::BasicBlock> merge_block =
+      MakeUnique<opt::BasicBlock>(MakeUnique<opt::Instruction>(opt::Instruction(
+          ir_context, SpvOpLabel, 0, message_.merge_label_fresh_id(),
+          opt::Instruction::OperandList())));
+  merge_block->SetParent(enclosing_function);
+
+  // Get the maps from the protobuf.
+  std::map<uint32_t, uint32_t> original_label_to_duplicate_label =
+      fuzzerutil::RepeatedUInt32PairToMap(
+          message_.original_label_to_duplicate_label());
+
+  std::map<uint32_t, uint32_t> original_id_to_duplicate_id =
+      fuzzerutil::RepeatedUInt32PairToMap(
+          message_.original_id_to_duplicate_id());
+
+  std::map<uint32_t, uint32_t> original_id_to_phi_id =
+      fuzzerutil::RepeatedUInt32PairToMap(message_.original_id_to_phi_id());
+
+  // Before adding duplicate blocks, we need to update the OpPhi instructions in
+  // the successors of the |exit_block|. We know that the execution of the
+  // transformed region will end in |merge_block|. Hence, we need to change all
+  // occurrences of the label id of the |exit_block| to the label id of the
+  // |merge_block|.
+  exit_block->ForEachSuccessorLabel([this, ir_context](uint32_t label_id) {
+    auto block = ir_context->cfg()->block(label_id);
+    for (auto& instr : *block) {
+      if (instr.opcode() == SpvOpPhi) {
+        instr.ForEachId([this](uint32_t* id) {
+          if (*id == message_.exit_block_id()) {
+            *id = message_.merge_label_fresh_id();
+          }
+        });
+      }
+    }
+  });
+
+  // Similarly, we need to update OpPhi instructions in the |entry_block|. We
+  // know that it has only one predecessor, since the region is single-entry,
```

Comment:
(As written in another comment) does being a single-entry, single-exit region imply that constructs, and their correponding merge or continue constructs, are wholly within or wholly ouside of the region? Would it be better to specify?